### ISSUE TITLE
[Question] How to log episode reward after every episode (without shift)?

### ISSUE THREAD

**Issue by jonasreiher** [DLR-RM/stable-baselines3#1300]
### ❓ Question

The values logged as `rollout/ep_rew_mean` are smoothed over 100 episodes. I would like to additionally report the total episode reward of each episode itself. (Related: #313, #400)

Now, in a callback I can access the `self.model.ep_info_buffer` which stores the 100 moving window rewards. Basically, I would like to access the last value here, or in my case the last 10 values since I'm running 10 environments in parallel.

However, when I notice the end of an episode in my callback by observing `self.locals['dones']`, this buffer does not yet contain the latest episode reward. Only in the next step the buffer is filled but then the logger has dumped its logs already. So if the first episode is over, I cannot log anything. If the second episode is over, I can only log the reward of the first one and so on.

How can I get rid of this shift?
Is there any other way besides `self.model.ep_info_buffer` where the episode reward can be accessed as soon as an episode is 'done'?

---

My callback (for multiple parallel environments) currently looks like this but logging is shifted. First counting dones/episodes and then checking if I should log results is trying to access an empty deque.
```python
class EpisodeRewardCallback(BaseCallback):

    def _on_training_start(self):
        self.locals['episodes'] = 0
        self.locals['prev_dones'] = 0

    def _on_step(self) -> bool:
        if (self.locals['prev_dones'] > 0):
            if (self.locals['episodes'] % self.training_env.num_envs) == 0:
                mean_ep_reward = np.mean([self.model.ep_info_buffer[-i-1]['r']
                                          for i in range(self.training_env.num_envs)])
                self.logger.record('episode/reward', mean_ep_reward)

        self.locals['prev_dones'] = self.locals['dones'].sum()
        self.locals['episodes'] += self.locals['prev_dones']

        return True
```

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
> However, when I notice the end of an episode in my callback by observing self.locals['dones']

it looks like you are observing the wrong variable, `self.locals["infos"]` is the one you should monitor:
https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/base_class.py#L423-L439

and

https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/on_policy_algorithm.py#L261-L263

You might have a look at:
https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/rl_zoo3/callbacks.py#L208-L239

https://github.com/DLR-RM/rl-baselines3-zoo/pull/216

**Comment by jonasreiher** [comment]
> `self.locals["infos"]` is the one you should monitor

`self.locals['infos']` and `self.locals['dones']` always seem to go in sync. If I see a `'TimeLimit.truncated': True` or `'is_success': True` there, I also see it in `self.locals['dones']`, so which one I'm observing shouldn't matter.

The `model.ep_info_buffer` is only updated afterwards, though. And that's where I want to access the reward/return and episode length. So a shift of one step is unavoidable this way.

I suppose your suggestion may have been to read episodic length and return from `self.locals['infos']` as well, as the `RawStatisticsCallback` seems to do. However, the info dictionaries _never_ contain an `'episode'` key for me. Only `'is_success'`, `'TimeLimit.truncated'` and `'terminal_observation'` for 'done' episodes.

So the `RawStatisticsCallback` never logs anything here. I'm using SAC and a custom observation wrapper for my environments. Could this be a problem somehow?

Either way, for now I'm sticking with the solution described above.

**Comment by jonasreiher** [comment]
**Correction**

The missing `'episode'` key was due to me using a `VecMonitor` wrapper on the `SubprovVecEnv` instead of individual `Monitor` wrapper per environment.

Even though [this line](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/vec_env/vec_monitor.py#L88) exists, this part does not show up in the callback's `locals['infos']`.

**Comment by araffin** [comment]
> Even though [this line](https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/vec_env/vec_monitor.py#L88) exists, this part does not show up in the callback's locals['infos'] ...

looks like a bug, I will try to reproduce the issue, thanks for reporting.

> So, keeping in mind this limitation, your suggestion is entirely valid and a good alternative to my solution, thanks!

good to hear =)

**Comment by araffin** [comment]
> The missing 'episode' key was due to me using a VecMonitor wrapper on the SubprovVecEnv instead of individual Monitor wrappers per environment.

I couldn't reproduce the issue...

```python
import gym

from gym import spaces
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv
from stable_baselines3.common.vec_env import VecMonitor
import numpy as np


class CustomEnv(gym.Env):
    def __init__(self, action_dim=2, state_dim=13):
        self.action_space = spaces.Box(low=-1, high=1, shape=(action_dim,))
        self.observation_space = spaces.Box(
            low=-np.inf,
            high=np.inf,
            shape=(state_dim,),
        )

    def reset(self):
        return self.observation_space.sample()

    def step(self, action):
        return self.observation_space.sample(), 0, np.random.rand() > 0.8, {}


def make_env():
    return CustomEnv


if __name__ == "__main__":
    # env = VecMonitor(DummyVecEnv([make_env() for _ in range(3)]))
    env = VecMonitor(SubprocVecEnv([make_env() for _ in range(3)]))

    for _ in range(10):
        obs = env.reset()
        for _ in range(100):
            action = [env.action_space.sample() for _ in range(env.num_envs)]
            obs, reward, done, info = env.step(action)
            if "episode" in info[0]:
                print(info)

```

**Comment by amanuelanteneh** [comment]
Hi @araffin, I looked at the code you linked above and tried it out but it seems to still be using the timestep as the x axis. Is there a straightforward way to simply plot the final reward of each episode as a function of the episode number?

**Comment by araffin** [comment]
> Hi @araffin, I looked at the code you linked above and tried it out but it seems to still be using the timestep as the x axis. Is there a straightforward way to simply plot the final reward of each episode as a function of the episode number?

if you don't need live update, this is what the "plot_train" script can do for you in the zoo.
Otherwise, you can set and increment an episode counter, that should do the trick.

**Comment by amanuelanteneh** [comment]
thanks @araffin that did the trick!
